patches/server/1055-Add-listing-API-for-PlayerProfile.patch

```diff
@@ -0,0 +1,133 @@
+     // Paper end - Add Listing API for Player
+ 
++    // Paper start - Add Listing API for PlayerProfile
++    @Override
++    public void listProfile(@NotNull com.destroystokyo.paper.profile.PlayerProfile profile, int latency, GameMode gameMode, @Nullable net.kyori.adventure.text.Component displayName) {
++        Preconditions.checkNotNull(profile, "profile cannot be null");
++        Preconditions.checkArgument(isOfflineProfile(profile), "using the profile of an online player is not allowed");
++        if (this.getHandle().connection == null) return;
++
++        GameType type = GameType.byId(gameMode.getValue());
++        net.minecraft.network.chat.Component name = io.papermc.paper.adventure.PaperAdventure.asVanilla(displayName);
++        com.mojang.authlib.GameProfile gameProfile = ((com.destroystokyo.paper.profile.CraftPlayerProfile) profile).buildGameProfile();
++        ClientboundPlayerInfoUpdatePacket packet = ClientboundPlayerInfoUpdatePacket.addAndList(gameProfile, latency, type, name);
++        this.getHandle().connection.send(packet);
++    }
++
++    @Override
++    public void updateListedProfileLatency(@NotNull com.destroystokyo.paper.profile.PlayerProfile profile, int latency) {
++        Preconditions.checkNotNull(profile, "profile cannot be null");
++        Preconditions.checkArgument(isOfflineProfile(profile), "using the profile of an online player is not allowed");
++        if (this.getHandle().connection == null) return;
++
++        this.getHandle().connection.send(ClientboundPlayerInfoUpdatePacket.updateLatency(profile.getId(), latency));
++    }
++
++    @Override
++    public void updateListedProfileGameMode(@NotNull com.destroystokyo.paper.profile.PlayerProfile profile, GameMode gameMode) {
++        Preconditions.checkNotNull(profile, "profile cannot be null");
++        Preconditions.checkArgument(isOfflineProfile(profile), "using the profile of an online player is not allowed");
++        if (this.getHandle().connection == null) return;
++
++        this.getHandle().connection.send(ClientboundPlayerInfoUpdatePacket.updateGameMode(profile.getId(), GameType.byId(gameMode.getValue())));
++    }
++
++    @Override
++    public void updateListedProfileDisplayName(@NotNull com.destroystokyo.paper.profile.PlayerProfile profile, @Nullable net.kyori.adventure.text.Component displayName) {
++        Preconditions.checkNotNull(profile, "profile cannot be null");
++        Preconditions.checkArgument(isOfflineProfile(profile), "using the profile of an online player is not allowed");
++        if (this.getHandle().connection == null) return;
++
++        this.getHandle().connection.send(ClientboundPlayerInfoUpdatePacket.updateDisplayName(profile.getId(), io.papermc.paper.adventure.PaperAdventure.asVanilla(displayName)));
++    }
++
++    @Override
++    public void unlistProfile(@NotNull com.destroystokyo.paper.profile.PlayerProfile profile) {
++        Preconditions.checkNotNull(profile, "profile cannot be null");
```

Comment:
We prefer using `checkArgument(profile != null)` because that will throw an `IllegalArgumentException` instead of an `NullPointerException`.